prov/gni/include/gnix_freelist.h

```diff
@@ -123,6 +126,12 @@ static inline int _gnix_fl_alloc(struct dlist_entry **e, struct gnix_freelist *f
         fastlock_acquire(&fl->lock);
 
     if (dlist_empty(&fl->freelist)) {
+
+        if (fl->refill_size == 0) {
+            ret = -FI_ECANCELED;
```

Comment:
I think that is a bit confusing. -FI_ENOMEM suggests that it couldn't allocate the memory or free entries necessary to satisfy the request. However, this error condition is related to whether or not it is permitted to satisfy the request. If the refill size is zero, it is not allowed to generate new entries and cannot return one.